server/events/vcs/github_client.go

```diff
@@ -307,6 +309,40 @@ func (g *GithubClient) PullIsMergeable(repo models.Repo, pull models.PullRequest
 	return true, nil
 }
 
+// Check if the Pull Request is locked with :lock emoji
+func (g *GithubClient) PullIsLocked(repo models.Repo, pull models.PullRequest) (bool, error) {
+	statuses, err := g.getRepoStatuses(repo, pull)
```

Comment:
Couple problems with this approach:

1) The call to this happens for every project. This is could end up really impacting our rate limits/performance in the long run (the same can be said for getting the approved status so not trying to build on that)
2) We already make the call to getRepoStatuses when we fetch [mergeability](https://github.com/lyft/atlantis/blob/release-v0.17.1-lyft.1/server/events/vcs/github_client.go#L313) so it makes sense to try and leverage this.


⚡  idea:

Create an interface which is responsible for fetching approval, mergeability and lock status:

```go
type PullStatus struct {
  approved bool
  mergeable bool
  sqLocked bool
}

type PullStatusFetcher interface {
    fetchPullStatus(...) (PullStatus, error)
}

type SQBasedPullStatusFetcher struct {
   approvedPullChecker PullApprovalChecker
   githubClient. GithubClient
}

func fetchPullStatus(...) (PullStatus, error) {
    // call github method to fetch statuses
    // populate mergeability flag
    // populated locked flag.
    ...
}

```

And then this method would be called once basically replacing the existing mergeability [call](https://github.com/lyft/atlantis/blob/release-v0.17.1-lyft.1/server/events/apply_command_runner.go#L101) and then propagate the pull status

Reply:
I've made changes to the architecture accordingly. Lmk if we can further improve the implementation.